core/src/main/java/org/frankframework/pipes/AsyncSenderWithListenerPipe.java

```diff
@@ -0,0 +1,290 @@
+@ElementType(ElementTypes.ENDPOINT)
+public class AsyncSenderWithListenerPipe<M> extends MessageSendingPipe {
+	private @Getter ICorrelatedPullingListener<M> listener = null;
+	private @Setter ListenerProcessor<M> listenerProcessor;
+
+	private @Getter String correlationIDStyleSheet;
+	private @Getter String correlationIDXPath;
+	private @Getter String correlationIDNamespaceDefs;
+	private @Getter String correlationIDSessionKey = null;
+	private @Getter String labelStyleSheet;
+	private @Getter String labelXPath;
+	private @Getter String labelNamespaceDefs;
+	private @Getter String auditTrailSessionKey = null;
+	private @Getter String auditTrailXPath;
+	private @Getter String auditTrailNamespaceDefs;
+	private @Getter boolean useInputForExtract = true;
+
+	private TransformerPool auditTrailTp = null;
+	private TransformerPool correlationIDTp = null;
+	private TransformerPool labelTp = null;
+
+	@Override
+	public void configure() throws ConfigurationException {
+		if (listener == null) {
+			// We use the Mandatory annotation, but just in case.
+			throw new ConfigurationException("Using a listener is mandatory");
+		}
+
+		if (getSender().isSynchronous()) {
+			throw new ConfigurationException("cannot have listener with synchronous sender");
+		}
+		try {
+			listener.configure();
+		} catch (ConfigurationException e) {
+			throw new ConfigurationException("while configuring listener",e);
+		}
+		if (listener instanceof HasPhysicalDestination dest) {
+			log.debug("has listener on {}", dest::getPhysicalDestinationName);
+		}
+
+		super.configure();
+
+		if (StringUtils.isNotEmpty(getAuditTrailXPath())) {
+			auditTrailTp = TransformerPool.configureTransformer(this, getAuditTrailNamespaceDefs(), getAuditTrailXPath(), null, OutputType.TEXT, false, null);
+		}
+		if (StringUtils.isNotEmpty(getCorrelationIDXPath()) || StringUtils.isNotEmpty(getCorrelationIDStyleSheet())) {
+			correlationIDTp = TransformerPool.configureTransformer(this, getCorrelationIDNamespaceDefs(), getCorrelationIDXPath(), getCorrelationIDStyleSheet(), OutputType.TEXT, false, null);
+		}
+		if (StringUtils.isNotEmpty(getLabelXPath()) || StringUtils.isNotEmpty(getLabelStyleSheet())) {
+			labelTp = TransformerPool.configureTransformer(this, getLabelNamespaceDefs(), getLabelXPath(), getLabelStyleSheet(), OutputType.TEXT, false, null);
+		}
+	}
+
+	@Override
+	protected void propagateName() {
+		super.propagateName();
+
+		if (StringUtils.isEmpty(listener.getName())) {
+			listener.setName(getName() + "-replylistener");
+		}
+	}
+
+	@Override
+	protected String doLogToMessageLog(final Message input, final PipeLineSession session, final Message originalMessage, final String messageID, String correlationID) throws SenderException {
+		try {
+			String messageTrail="no audit trail";
+			if (auditTrailTp!=null) {
+				if (isUseInputForExtract()){
+					messageTrail=auditTrailTp.transform(originalMessage);
+				} else {
+					messageTrail=auditTrailTp.transform(input);
+				}
```

Comment:
Dit is verwarrend: `if (useInput == true) then use something different, else use input`

Ik snap dat `useInputForExtract` een legacy configuration name is dus misschien kan de method-parameter worden hernoemd?

Reply:
Poeh, laten we hier een volgende keer naar kijken, dit zit erg verwoven in de code en vaak is de input  gelijk aan de originalMessage...